docs/enhacement/20201211-nodepool_uniteddeployment.md

```diff
@@ -0,0 +1,365 @@
+type StatefulSetTemplateSpec struct {
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	Spec              appsv1.StatefulSetSpec `json:"spec"`
+}
+
+// DeploymentTemplateSpec defines the pool template of Deployment.
+type DeploymentTemplateSpec struct {
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	Spec              appsv1.DeploymentSpec `json:"spec"`
+}
+
+// UnitedDeploymentUpdateStrategy defines the update performance
+// when template of UnitedDeployment is changed.
+// Only one of its members may be specified.
+type UnitedDeploymentUpdateStrategy struct {
+	// Includes all of the parameters a statefulset update strategy needs.
+	// +optional
+	StatefulSetUpdateStrategy *StatefulSetUpdateStrategy `json:"statefulSetUpdateStrategy,omitempty"`
+
+	// Includes all of the parameters a deployment update strategy needs.
+	// +optional
+	DeploymentUpdateStrategy *DeploymentUpdateStrategy `json:"deploymentUpdateStrategy,omitempty"`
+}
+
+// StatefulSetUpdateStrategy is a update strategy which allows users to control the update progress
+// by providing the partition of each pool.
+type StatefulSetUpdateStrategy struct {
+	// Indicates strategy of statefulset for each pool.
+	// +optional
+	Strategies map[string]appsv1.StatefulSetUpdateStrategy `json:"strategies,omitempty"`
+}
+
+// DeploymentUpdateStrategy is a update strategy
+type DeploymentUpdateStrategy struct {
+	// Indicates strategy of deployment for each pool.
+	// +optional
+	Strategies map[string]appsv1.DeploymentStrategy `json:"strategies,omitempty"`
+}
+
+// Topology defines the spread detail of each pool under UnitedDeployment.
+// A UnitedDeployment manages multiple homogeneous workloads which are called pool.
+// Each of pools under the UnitedDeployment is described in Topology.
+type Topology struct {
+	// Contains the details of each pool. Each element in this array represents one pool
+	// which will be provisioned and managed by UnitedDeployment.
+	// +optional
+	Pools []Pool `json:"pools,omitempty"`
+}
+
+// Pool defines the detail of a pool.
+type Pool struct {
+	// Indicates pool name as a DNS_LABEL, which will be used to generate
+	// pool workload name prefix in the format '<deployment-name>-<pool-name>-'.
+	// Name should be unique between all of the pools under one UnitedDeployment.
+	// Name is NodePool Name
+	Name string `json:"name"`
+
+	// Indicates the node selector to form the pool. Depending on the node selector,
+	// pods provisioned could be distributed across multiple groups of nodes.
+	// A pool's nodeSelectorTerm is not allowed to be updated.
+	// +optional
+	NodeSelectorTerm corev1.NodeSelectorTerm `json:"nodeSelectorTerm,omitempty"`
+
+	// Indicates the tolerations the pods under this pool have.
+	// A pool's tolerations is not allowed to be updated.
+	// +optional
+	Tolerations []corev1.Toleration `json:"tolerations,omitempty"`
+
+	// Indicates the number of the pod to be created under this pool.
+	// +optional
```

Comment:
It looks that the Replica field needs to be a required field.

Reply:
done